### PR TITLE
Fixups for llvmlite 0.23/llvm 6

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

### CODE REVIEW

**Comment by stuartarchibald** [comment]
Needs https://github.com/numba/llvmlite/pull/340

**Comment by stuartarchibald** [comment]
At present, the test: `<path>/numba/tests/npyufunc/test_parallel_ufunc_issues.py, line 119 in test_gil_reacquire_deadlock` is failing with a segmentation fault. This is reproducible by simply running:

```
python -m numba.runtests -m -b numba.tests.npyufunc
```

==7166== Process terminating with default action of signal 11 (SIGSEGV)
==7166==    at 0x4E454AB: raise (pt-raise.c:37)
==7166==    by 0x4E455DF: ??? (in /usr/lib64/libpthread-2.17.so)
==7166==    by 0xDEAFB30: ffi_closure_unix64_inner (in <path>/lib/libffi.so.6.0.4)
==7166==    by 0xDEB0027: ffi_closure_unix64 (in <path>/lib/libffi.so.6.0.4)
==7166==    by 0x1EFD5114: ???
==7166==    by 0x11EFD4FFF: ???
==7166==    by 0x166B57DF: ???

Simultaneously attaching a gdb session to all forked processes that are running under the test runner and allowing them to run in the background allows switching to the process that segfaults at the time of error. The following is some basic gdb output:

```
Program received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7f9a46ec8700 (LWP 2402)]
0x00007f9a54b0cb31 in ffi_closure_unix64_inner () from
<path>/lib/python3.6/lib-dynload/../../libffi.so.6
(gdb) bt
#0  0x00007f9a54b0cb31 in ffi_closure_unix64_inner () from
<path>/lib/python3.6/lib-dynload/../../libffi.so.6
#1  0x00007f9a54b0d028 in ffi_closure_unix64 () from
<path>/lib/python3.6/lib-dynload/../../libffi.so.6
#2  0x00007f9a46458115 in
numba::tests::npyufunc::test_parallel_ufunc_issues::TestParGUfuncIssues::test_gil_reacquire_deadlock::$3clocals$3e::foo$2434(int,
Array<int, 1, A, mutable, aligned>) ()
#3  0x00007f9a46455075 in
__gufunc__._ZN5numba5tests8npyufunc26test_parallel_ufunc_issues19TestParGUfuncIssues27test_gil_reacquire_deadlock12$3clocals$3e8foo$2434Ei5ArrayIiLi1E1A7mutable7alignedE
()
#4  0x00007f9a4a6d2110 in thread_worker () from
<path>/numba/npyufunc/workqueue.cpython-36m-x86_64-linux-gnu.so
#5  0x00007f9a5d3b8e25 in start_thread (arg=0x7f9a46ec8700) at
pthread_create.c:308
#6  0x00007f9a5d0e634d in clone () at
../sysdeps/unix/sysv/linux/x86_64/clone.S:113
(gdb) info register
rax            0x736e6920       1936615712
rbx            0x7f9a46ec7e40   140300591595072
rcx            0x0      0
rdx            0x7f9a46ec7d70   140300591594864
rsi            0x7f9a4ae20f90   140300658020240
rdi            0x69202c2a38697b20       7575103133083859744
rbp            0x7f9a46ec7d60   0x7f9a46ec7d60
rsp            0x7f9a46ec7cb0   0x7f9a46ec7cb0
r8             0x0      0
r9             0x21750  137040
r10            0x7f9a5d7e5080   140300970250368
r11            0x7f9a54b0cfe2   140300822564834
r12            0x55a879905aa2   94182082370210
r13            0x0      0
r14            0x7f9a46ec7ed0   140300591595216
r15            0x0      0
rip            0x7f9a54b0cb31   0x7f9a54b0cb31 <ffi_closure_unix64_inner+81>
eflags         0x10246  [ PF ZF IF RF ]
cs             0x33     51
ss             0x2b     43
ds             0x0      0
es             0x0      0
fs             0x0      0
gs             0x0      0
(gdb) info threads
  Id   Target Id         Frame
* 9    Thread 0x7f9a46ec8700 (LWP 2402) "python" 0x00007f9a54b0cb31 in
ffi_closure_unix64_inner () from
<path>/lib/python3.6/lib-dynload/../../libffi.so.6
  8    Thread 0x7f9a476c9700 (LWP 2396) "python"
pthread_cond_wait@@GLIBC_2.3.2 () at
../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  7    Thread 0x7f9a47eca700 (LWP 2393) "python"
pthread_cond_wait@@GLIBC_2.3.2 () at
../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  6    Thread 0x7f9a486cb700 (LWP 2392) "python"
pthread_cond_wait@@GLIBC_2.3.2 () at
../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  5    Thread 0x7f9a48ecc700 (LWP 2390) "python"
pthread_cond_wait@@GLIBC_2.3.2 () at
../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  4    Thread 0x7f9a496cd700 (LWP 2389) "python"
pthread_cond_wait@@GLIBC_2.3.2 () at
../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  3    Thread 0x7f9a49ece700 (LWP 2386) "python"
pthread_cond_wait@@GLIBC_2.3.2 () at
../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  2    Thread 0x7f9a4a6cf700 (LWP 2384) "python"
pthread_cond_wait@@GLIBC_2.3.2 () at
../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  1    Thread 0x7f9a5d7d0740 (LWP 2222) "python"
pthread_cond_wait@@GLIBC_2.3.2 () at
../nptl/sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
```

I think this points to possible corruption, running under valgrind has some worrying items:

```
==7169==  Address 0x1171c020 is 8 bytes after a block of size 8 free'd
==7169==    at 0x4C2B18D: operator delete(void*) (vg_replace_malloc.c:576)
==7169==    by 0xE9FF02B: void std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> >::_M_range_insert<std::reverse_iterator<__gnu_cxx::__normal_iterator<llvm::MachineBasicBlock**, std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> > > > >(__gnu_cxx::__normal_iterator<llvm::MachineBasicBlock**, std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> > >, std::reverse_iterator<__gnu_cxx::__normal_iterator<llvm::MachineBasicBlock**, 
std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> > > >, std::reverse_iterator<__gnu_cxx::__normal_iterator<llvm::MachineBasicBlock**, std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> > > >, std::forward_iterator_tag) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE9FF64B: llvm::LiveVariables::MarkVirtRegAliveInBlock(llvm::LiveVariables::VarInfo&, llvm::MachineBasicBlock*, llvm::MachineBasicBlock*, std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> >&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE9FF83E: llvm::LiveVariables::MarkVirtRegAliveInBlock(llvm::LiveVariables::VarInfo&, llvm::MachineBasicBlock*, llvm::MachineBasicBlock*) (in <path>/lib/python3.6/site-packages/llvmlite/binding/lib
llvmlite.so)
==7169==    by 0xEA014BA: llvm::LiveVariables::HandleVirtRegUse(unsigned int, llvm::MachineBasicBlock*, llvm::MachineInstr&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xEA06914: llvm::LiveVariables::runOnInstr(llvm::MachineInstr&, llvm::SmallVectorImpl<unsigned int>&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xEA06F9D: llvm::LiveVariables::runOnBlock(llvm::MachineBasicBlock*, unsigned int) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xEA084D0: llvm::LiveVariables::runOnMachineFunction(llvm::MachineFunction&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xEA6E384: llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF70D450: llvm::FPPassManager::runOnFunction(llvm::Function&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF70D4B9: llvm::FPPassManager::runOnModule(llvm::Module&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF70CB2E: llvm::legacy::PassManagerImpl::run(llvm::Module&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==  Block was alloc'd at
==7169==    at 0x4C2A203: operator new(unsigned long) (vg_replace_malloc.c:334)
==7169==    by 0xE9FEE00: void std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> >::_M_range_insert
<std::reverse_iterator<__gnu_cxx::__normal_iterator<llvm::MachineBasicBlock**, std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> > > > >(__gnu_cxx::__normal_iterator<llvm::MachineBasicBlock**, std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> > >, std::reverse_iterator<__gnu_cxx::__normal_iterator<llvm::MachineBasicBlock**, 
std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> > > >, std::reverse_iterator<__gnu_cxx::__normal_
iterator<llvm::MachineBasicBlock**, std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> > > >, std::forward_iterator_tag) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE9FF64B: llvm::LiveVariables::MarkVirtRegAliveInBlock(llvm::LiveVariables::VarInfo&, llvm::MachineBasicBlock*, llvm::MachineBasicBlock*, std::vector<llvm::MachineBasicBlock*, std::allocator<llvm::MachineBasicBlock*> >&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE9FF80F: llvm::LiveVariables::MarkVirtRegAliveInBlock(llvm::LiveVariables::VarInfo&, llvm::MachineBasicBlock*, llvm::MachineBasicBlock*) (in <path>/lib/python3.6/site-packages/llvmlite/binding/lib
llvmlite.so)
==7169==    by 0xEA014BA: llvm::LiveVariables::HandleVirtRegUse(unsigned int, llvm::MachineBasicBlock*, llvm::MachineInstr&) (in
 <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xEA06914: llvm::LiveVariables::runOnInstr(llvm::MachineInstr&, llvm::SmallVectorImpl<unsigned int>&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xEA06F9D: llvm::LiveVariables::runOnBlock(llvm::MachineBasicBlock*, unsigned int) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xEA084D0: llvm::LiveVariables::runOnMachineFunction(llvm::MachineFunction&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xEA6E384: llvm::MachineFunctionPass::runOnFunction(llvm::Function&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF70D450: llvm::FPPassManager::runOnFunction(llvm::Function&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF70D4B9: llvm::FPPassManager::runOnModule(llvm::Module&) (in <path>/
lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF70CB2E: llvm::legacy::PassManagerImpl::run(llvm::Module&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)

```
and
```
==7169== Invalid read of size 4
==7169==    at 0x1EFD8D: PyMem_Realloc (in <path>/bin/python3.6)
==7169==    by 0x2640C0: listpop (in <path>/bin/python3.6)
==7169==    by 0x216700: _PyCFunction_FastCallDict (in <path>/bin/python3.6)
==7169==    by 0x2A410B: call_function (in <path>/bin/python3.6)
==7169==    by 0x2C8419: _PyEval_EvalFrameDefault (in <path>/bin/python3.6)
==7169==    by 0x29D0A5: _PyEval_EvalCodeWithName (in <path>/bin/python3.6)
==7169==    by 0x29E0E0: fast_function (in <path>/bin/python3.6)
==7169==    by 0x2A41E4: call_function (in <path>/bin/python3.6)
==7169==    by 0x2C8419: _PyEval_EvalFrameDefault (in <path>/bin/python3.6)
==7169==    by 0x29DEAA: fast_function (in <path>/bin/python3.6)
==7169==    by 0x2A41E4: call_function (in <path>/bin/python3.6)
==7169==    by 0x2C8419: _PyEval_EvalFrameDefault (in <path>/bin/python3.6)
==7169==  Address 0x15612020 is 176 bytes inside a block of size 224 free'd
==7169==    at 0x4C2B18D: operator delete(void*) (vg_replace_malloc.c:576)
==7169==    by 0xF5A1F5E: llvm::MCSection::~MCSection() (in <path>/lib/python3.6/site
-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF5787BC: llvm::MCContext::reset() (in <path>/lib/python3.6/site-pack
ages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xEA9254C: llvm::MachineModuleInfo::doFinalization(llvm::Module&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF70CCD0: llvm::legacy::PassManagerImpl::run(llvm::Module&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE8E8D72: llvm::MCJIT::emitObject(llvm::Module*) (in <path>/lib/pytho
n3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE8E9772: llvm::MCJIT::generateCodeForModule(llvm::Module*) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE8E4420: llvm::MCJIT::finalizeObject() (in <path>/lib/python3.6/site
-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xDEAFEBF: ffi_call_unix64 (in <path>/lib/libffi.so.6.0.4)
==7169==    by 0xDEAF87C: ffi_call (in <path>/lib/libffi.so.6.0.4)
==7169==    by 0xDC98DED: _ctypes_callproc (in <path>/lib/python3.6/lib-dynload/_ctyp
es.cpython-36m-x86_64-linux-gnu.so)
==7169==    by 0xDC99747: PyCFuncPtr_call (in <path>/lib/python3.6/lib-dynload/_ctype
s.cpython-36m-x86_64-linux-gnu.so)
==7169==  Block was alloc'd at
==7169==    at 0x4C2A203: operator new(unsigned long) (vg_replace_malloc.c:334)
==7169==    by 0xF57CF2F: llvm::MCContext::createELFSectionImpl(llvm::StringRef, unsigned int, unsigned int, llvm::SectionKind, 
unsigned int, llvm::MCSymbolELF const*, unsigned int, llvm::MCSymbolELF const*) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF57DB85: llvm::MCContext::getELFSection(llvm::Twine const&, unsigned int, unsigned int, unsigned int, llvm::MCS
ymbolELF const*, unsigned int, llvm::MCSymbolELF const*) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF57F127: llvm::MCContext::getELFSection(llvm::Twine const&, unsigned int, unsigned int, unsigned int, llvm::Twi
ne const&, unsigned int, llvm::MCSymbolELF const*) (in <path>/lib/python3.6/site-pack
ages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF59C11B: llvm::MCObjectFileInfo::initELFMCObjectFileInfo(llvm::Triple const&, bool) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE3F03B1: llvm::X86LinuxNaClTargetObjectFile::Initialize(llvm::MCContext&, llvm::TargetMachine const&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE80CB51: llvm::AsmPrinter::doInitialization(llvm::Module&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF70274E: llvm::FPPassManager::doInitialization(llvm::Module&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xF70C982: llvm::legacy::PassManagerImpl::run(llvm::Module&) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE8E8D72: llvm::MCJIT::emitObject(llvm::Module*) (in <path>/lib/pytho
n3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE8E9772: llvm::MCJIT::generateCodeForModule(llvm::Module*) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE8E4420: llvm::MCJIT::finalizeObject() (in <path>/lib/python3.6/site
-packages/llvmlite/binding/libllvmlite.so)
```
and then the segfault
```
==7169== Conditional jump or move depends on uninitialised value(s)
==7169==    at 0x1E7FA3: _PyObject_Free (in <path>/bin/python3.6)
==7169==    by 0x29ED8B: PyEval_EvalCodeEx (in <path>/bin/python3.6)
==7169==    by 0x29FA13: function_call (in <path>/bin/python3.6)
==7169==    by 0x2165CD: PyObject_Call (in <path>/bin/python3.6)
==7169==    by 0xDC988C0: closure_fcn (in <path>/lib/python3.6/lib-dynload/_ctypes.cpython-36m-x86_64-linux-gnu.so)
==7169==    by 0xDEAFC5F: ffi_closure_unix64_inner (in <path>/lib/libffi.so.6.0.4)
==7169==    by 0xDEB0027: ffi_closure_unix64 (in <path>/lib/libffi.so.6.0.4)
==7169==    by 0xE38D03D: LLVMPYObjectCache::notifyObjectCompiled(llvm::Module const*, llvm::MemoryBufferRef) (executionengine.cpp:199)
==7169==    by 0xE8E8E9D: llvm::MCJIT::emitObject(llvm::Module*) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE8E9772: llvm::MCJIT::generateCodeForModule(llvm::Module*) (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xE8E4420: llvm::MCJIT::finalizeObject() (in <path>/lib/python3.6/site-packages/llvmlite/binding/libllvmlite.so)
==7169==    by 0xDEAFEBF: ffi_call_unix64 (in <path>/lib/libffi.so.6.0.4)
==7169== 
.Fatal Python error: Segmentation fault
```

**Comment by stuartarchibald** [comment]
The above segfault reproduces on the same machine using the shipped Numba 0.37. Seems like this failure is not related to moving to LLVM 6.

**Comment by stuartarchibald** [comment]
Update:
* llvmlite needs an adjustment to the code model handing to accommodate a `default` mode (failing on 32bit due to this).
* OSX is failing tests to do with `pycc` as there's a clang-4.0 linker error (object file being linked for newer OSX version that being linked).
* Windows 64bit and Linux x86_64 are fine.

**Comment by stuartarchibald** [comment]
The only known remaining issue is on OSX and in AOT testing. (With thanks to @mingwandroid for assistance in debugging/thoughts!).

All OSX builds fail with some variation on:
```
ld: illegal text-relocation to '_puts' in <OSX SDK PATH>/usr/lib/libSystem.tbd from '_._pycc_method_div' in <tmpdir>/pycc_test_simple.cpython-36m-darwin.o for architecture x86_64
clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
ld: warning: -pie being ignored. It is only used when linking a main executable
ld: warning: -read_only_relocs cannot be used with x86_64
ld: warning: object file (<tmpdir>/pycc_test_simple.cpython-36m-darwin.o) was built for new
er OSX version (10.11) than being linked (10.9)
ld: illegal text-relocation to '_puts' in <OSX SDK PATH>/usr/lib/libSystem.tbd from '_._pycc_method_div' in <tmpdir>/pycc_test_simple.cpython-36m-darwin.o for architecture x86_64
clang-4.0: error: linker command failed with exit code 1 (use -v to see invocation)
```
The variation present between tests is found in the relocated symbol, in the case above it is `_puts` but could just as easily be some symbol from the Numba NRT or elsewhere (juggling code generation/compiled code moves the relocations about such that different symbols are present first in the relocations).

A trivial sample to trigger this problem is:
```python
from numba.pycc import CC

cc = CC('amodule')

@cc.export('function1', 'void()')
def function1():
        pass

if __name__ == "__main__":
        cc.compile()
```
Inspecting the objects that clang is complaining about shows relocations like:
```
000000000000017a X86_64_RELOC_UNSIGNED _puts
```

On LLVM 5 the generated code for e.g. a `puts` call looked like:
```asm
leaq    "_.const.Fatal error: missing _dynfunc.Closure.11"(%rip), %rdi
callq    _puts
.cfi_endproc
```
i.e. the call is directly to `_puts`. Whereas on LLVM 6, the same functionality is written as:
```asm
movabsq $"_.const.Fatal error: missing _dynfunc.Closure", %rdi
movabsq $_puts, %rax
callq   *%rax
.cfi_endproc
```
i.e. load to register then call from register. This pattern ends up with a relocation record for the `.text` section for `_puts` and it is to this which `clang 4` objects.

There is a flag in `clang` to permit relocations from read only sections (`-read_only_relocs suppress`) but this gives `ld: warning: -read_only_relocs cannot be used with x86_64` and it therefore has no effect.

Other observations:
 * Linux generates the second pattern (load to register then call) on both LLVM 5 and LLVM 6. Under LLVM 5 a relocation for `puts` is in the relocation records for `.text` as:
```
<OFFSET> R_X86_64_PLT32 puts-0x0000000000000004
```
whereas under LLVM 6 the `PLT32 `is supplanted with `64`
```
<OFFSET> R_X86_64_64 puts
```
 * We wondered if this was `retpoline` related, but with/without setting `-mretpoline/-mno-retpoline` during codegen the result was the same.

**Comment by stuartarchibald** [comment]
Root cause appears to be setting the default code model to be based on machine bitness, this is fine for MCJIT but AOT compilation needs the small model. It also explains the difference in relocations and addressing.

File: numba/tests/test_parfors.py [hunk 1]
```diff
@@ -1596,9 +1596,10 @@ def test_vectorizer_fastmath_asm(self):
         # packed dtype=double add and sqrt instructions.
         def will_vectorize(A):
             n = len(A)
+            acc = 0
             for i in range(n):
-                A[i] += np.sqrt(i)
-            return A
+                acc += np.sqrt(i)
+            return acc
```
Comment (sklam) [inline numba/tests/test_parfors.py]: Is this change needed because the new LLVM6 will always vectorize the code, and, reduction-loop will only be vectorized if fastmath is ON?

Comment (stuartarchibald) [reply numba/tests/test_parfors.py]: I think that is the case. The LLVM 6 vectorizer seems to be able to vertically vectorize more loops. This change forces the need for associativity to be broken to vectorize, so should only work under `fastmath`.
